src/components/NumberInput/index.jsx

```diff
@@ -6,146 +6,131 @@ import classnames from 'classnames';
-    if (e.target.value === '') {
-      this.applyChange(e.target.value);
+  _onValueChanged({ target: { value: newValue } }) {
+    if (is.empty(newValue)) {
+      return this._sendCallbackWithNewValue(null);
     }
 
-    if (this.acceptedString.test(e.target.value)) {
-      this.setState({ value: e.target.value });
+    if (! REG_EXP_ACCEPTED_CHARS.test(newValue)) {
+      return;
+    }
 
-      if ((is.undefined(this.state.min) ||
-          Number(e.target.value, 10) >= this.state.min) &&
-          (is.undefined(this.state.max) ||
-          Number(e.target.value, 10) <= this.state.max)) {
-        this.applyChange(Number(e.target.value, 10));
-      }
+    const numberValue = Number(newValue, 10);
+    if ((is.undefined(this._min) || numberValue >= this._min) &&
+        (is.undefined(this._max) || numberValue <= this._max)) {
+      return this._sendCallbackWithNewValue(numberValue);
+    }
 
-      if (Number(e.target.value, 10) < this.state.min) {
-        this.applyChange(Number(this.state.min, 10));
-      }
+    if (numberValue < this._min) {
+      return this._sendCallbackWithNewValue(this._min);
+    }
 
-      if (Number(e.target.value, 10) > this.state.max) {
-        this.applyChange(Number(this.state.max, 10));
-      }
+    if (numberValue > this._max) {
+      return this._sendCallbackWithNewValue(this._max);
     }
   }
 
-  applyChange(value) {
-    this.setState({ value });
-    if (this.props.onChange) {
-      if (value === '') {
-        this.props.onChange({ target: {} });
-      } else {
-        this.props.onChange({ target: { value } });
-      }
+  _sendCallbackWithNewValue(value) {
+    if (is.function(this.props.onChange)) {
+      this.props.onChange(value);
     }
   }
 
+  _onFocus() {
+    this.setState({ focused: true });
+  }
+
+  _onBlur() {
+    this.setState({ focused: false });
+  }
+
   render() {
-    return <div className={styles.NumberInputContainer}>
-      {this.props.labelName ?
-        <span className={styles.NumberInputName}>
+    const classes = classnames(styles.NumberInput_inputContainer,
+        this.props.className,
+        { [styles.__focused]: this.state.focused });
+    return <div className={styles.NumberInput}>
+      {
+        this.props.labelName &&
+        <span className={styles.NumberInput_label}>
           {this.props.labelName}
-        </span> :
-        ''}
-      <div className={classnames(styles.NumberInput, this.props.className)}
-          style={this.props.customStyle} >
-        {
-          ! this.props.disabled &&
-          <div className={styles.upArrow} onClick={() => this.upArrowClickHandler()}>
-            <Icon id="chevron-small-up" />
-          </div>
-        }
+        </span>
+      }
+      <div className={classes}>
+        <input className={styles.NumberInput_inputContainer_input}
+            type="text"
+            value={this._value || this._defaultValue}
+            disabled={this.props.unwritable || this.props.disabled}
```

Comment:
`unwritable` is known as `readonly` in HTML lingo, but I think we can live with this.